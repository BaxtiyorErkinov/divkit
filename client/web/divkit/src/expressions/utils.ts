import type { EvalValue } from './eval';
import type { Node, Variable } from './ast';
import type { EvalTypes } from './eval';
import type { VariablesMap } from './eval';
import { walk } from './walk';
import { MAX_INT, MIN_INT } from './const';

export function valToInternal(val: EvalValue): EvalValue {
    if (val.type === 'url' || val.type === 'color') {
        return {
            type: 'string',
            value: val.value
        };
    }

    return val;
}

export function dateToString(date: Date): string {
    return date
        .toISOString()
        .replace('T', ' ')
        .replace(/\.\d{3}Z$/, '');
}

export function valToString(val: EvalValue): string {
    val = valToInternal(val);

    if (val.type === 'string') {
        return val.value;
    } else if (val.type === 'integer') {
        return String(val.value);
    } else if (val.type === 'number') {
        let res = String(val.value);

        if (!res.includes('.')) {
            if (res.includes('e')) {
                res = res.replace('e', '.0e');
            } else {
                res += '.0';
            }
        }

        res = res.replace(/e\+?/i, 'E');

        return res;
    } else if (val.type === 'boolean') {
        return val.value ? 'true' : 'false';
    } else if (val.type === 'datetime') {
        return dateToString(val.value);
    }

    throw new Error(`Unexpected type ${val.type}`);
}

export function valToPreview(val: EvalValue): string {
    let res = valToString(val);

    if (val.type === 'string') {
        res = "'" +
            res
                .replace(/\\/g, '\\\'')
                .replace(/'/g, '\\\'') +
            "'";
    }

    return res;
}

export function typeToString(type: EvalTypes): string {
    if (type === 'datetime') {
        return 'DateTime';
    }
    return type.charAt(0).toUpperCase() + type.substring(1);
}

const int32Array = new Int32Array(1);
export function roundInteger(val: number): number {
    int32Array[0] = val;
    return int32Array[0];
}

export function checkIntegerOverflow(val: number): void {
    if (val < MIN_INT || val > MAX_INT) {
        throw new Error('Integer overflow.');
    }
}

export function gatherVarsFromAst(ast: Node): string[] {
    const res: string[] = [];

    walk(ast, {
        Variable(node: Variable): void {
            res.push(node.id.name);
        }
    });

    return res;
}

export function evalError(msg: string, details: string): never {
    throw new Error(`Failed to evaluate [${msg}]. ${details}`);
}

export function containsUnsetVariables(ast: Node, variables: VariablesMap): boolean {
    let result = false;

    walk(ast, {
        Variable(node: Variable): void {
            if (!variables.has(node.id.name)) {
                result = true;
            }
        }
    });

    return result;
}
