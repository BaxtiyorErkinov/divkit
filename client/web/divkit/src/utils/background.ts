import type { AlignmentHorizontal, AlignmentVertical } from '../types/alignment';
import type { Background, GradientBackground, ImageBackground, SolidBackground } from '../types/background';
import type { ImageScale } from '../types/imageScale';
import type { MaybeMissing } from '../expressions/json';
import { correctColor } from './correctColor';
import { htmlFilter } from './htmlFilter';
import { Truthy } from './truthy';

export function getBackground(bgs: MaybeMissing<Background>[]): {
    color?: string;
    image?: string;
    size?: string;
    position?: string;
} {
    if (bgs.length === 1 && bgs[0].type === 'solid') {
        return solidUnique({
            bg: bgs[0]
        });
    }

    const res = bgs
        .map(bg => {
            if (bg.type === 'solid') {
                return solid({
                    bg
                });
            } else if (bg.type === 'gradient') {
                return gradient({
                    bg
                });
            } else if (bg.type === 'image') {
                return image({
                    bg
                });
            }
        })
        .filter(Truthy)
        .reverse()
        .reduce(function(acc, item) {
            acc.image.push(item.image);
            acc.size.push(item.size || 'auto');
            acc.position.push(item.pos || '50% 50%');

            return acc;
        }, {
            image: [] as string[],
            size: [] as string[],
            position: [] as string[]
        });

    return {
        image: res.image.join(','),
        size: res.size.join(','),
        position: res.position.join(',')
    };
}

function solid(opts: {
    bg: MaybeMissing<SolidBackground>;
}): {
    size: string | undefined;
    pos: string | undefined;
    image: string;
} {
    const color = correctColor(opts.bg.color || 'transparent');

    return {
        size: undefined,
        pos: undefined,
        image: `linear-gradient(to bottom,${color},${color})`
    };
}

function solidUnique(opts: {
    bg: MaybeMissing<SolidBackground>;
}): {
    color: string;
    size: string;
    position: string;
} {
    const color = correctColor(opts.bg.color || 'transparent');

    return {
        color,
        size: 'auto',
        position: '50% 50%'
    };
}

function gradient(opts: {
    bg: MaybeMissing<GradientBackground>;
}): {
    size: string | undefined;
    pos: string | undefined;
    image: string;
} | undefined {
    if (!Array.isArray(opts.bg?.colors)) {
        return;
    }

    const colors = opts.bg.colors.filter(Truthy) as string[];
    if (!colors.length) {
        return;
    }

    return {
        size: undefined,
        pos: undefined,
        image:
            'linear-gradient(' +
            (90 - Number(opts.bg.angle || 0) + 'deg') +
            ',' +
            colors
                .map(color => correctColor(color))
                .join(',') +
            ')'
    };
}

function image(opts: {
    bg: MaybeMissing<ImageBackground>;
}): {
    size: string | undefined;
    pos: string | undefined;
    image: string;
} | undefined {
    const image = opts.bg?.image_url;

    if (!image) {
        return;
    }

    return {
        size: imageSize(opts.bg.scale),
        pos: imagePos(opts.bg),
        image: 'url("' + htmlFilter(image) + '")'
    };
}

export function imageSize(scale?: ImageScale): string | undefined {
    if (scale === 'fill' || !scale) {
        return 'cover';
    } else if (scale === 'fit') {
        return 'contain';
    }
}

export function imagePos(obj: {
    content_alignment_horizontal?: AlignmentHorizontal;
    content_alignment_vertical?: AlignmentVertical;
}): string {
    let hpos: string;
    let vpos: string;

    if (obj.content_alignment_horizontal === 'left') {
        hpos = '0%';
    } else if (obj.content_alignment_horizontal === 'right') {
        hpos = '100%';
    } else {
        hpos = '50%';
    }

    if (obj.content_alignment_vertical === 'top') {
        vpos = '0%';
    } else if (obj.content_alignment_vertical === 'bottom') {
        vpos = '100%';
    } else {
        vpos = '50%';
    }

    return hpos + ' ' + vpos;
}
