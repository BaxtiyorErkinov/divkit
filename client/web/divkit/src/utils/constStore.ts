import type { Readable } from 'svelte/types/runtime/store';

// eslint-disable-next-line @typescript-eslint/no-empty-function
function voidUnsubscribe() {
}

export function constStore<T>(val: T): Readable<T> {
    return {
        subscribe(cb) {
            cb(val);
            return voidUnsubscribe;
        }
    };
}
