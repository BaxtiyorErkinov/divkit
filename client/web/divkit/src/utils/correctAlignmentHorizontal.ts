import type { AlignmentHorizontal } from '../types/alignment';

export function correctAlignmentHorizontal(
    orientation: string | undefined,
    defaultVal: AlignmentHorizontal
): AlignmentHorizontal {
    if (
        orientation === 'left' ||
        orientation === 'center' ||
        orientation === 'right'
    ) {
        return orientation;
    }

    return defaultVal;
}
