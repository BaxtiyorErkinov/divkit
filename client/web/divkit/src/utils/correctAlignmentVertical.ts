import type { AlignmentVertical } from '../types/alignment';

export function correctAlignmentVertical(
    orientation: string | undefined,
    defaultVal: AlignmentVertical
): AlignmentVertical {
    if (
        orientation === 'top' ||
        orientation === 'center' ||
        orientation === 'bottom'
    ) {
        return orientation;
    }

    return defaultVal;
}
