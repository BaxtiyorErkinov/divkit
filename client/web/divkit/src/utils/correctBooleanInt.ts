export function correctBooleanInt(val: number | undefined, defaultVal: boolean): boolean {
    if (val === 1 || val === 0) {
        return Boolean(val);
    }
    return defaultVal;
}
