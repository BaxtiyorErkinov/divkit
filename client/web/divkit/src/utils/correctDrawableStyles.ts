import type { Drawable } from '../types/slider';
import type { Style } from '../types/general';
import type { MaybeMissing } from '../expressions/json';
import { correctColor } from './correctColor';
import { pxToEm } from './pxToEm';

export interface DrawableStyle extends Style {
    width: number;
    height: number;
    borderRadius: number;
    background: string;
    boxShadow?: string;
}

export function correctDrawableStyle(
    drawable: MaybeMissing<Drawable> | undefined,
    defaultValue: DrawableStyle
): DrawableStyle {
    if (!drawable || !drawable.shape || !drawable.color) {
        return defaultValue;
    }

    const bg = correctColor(drawable.color);
    const width = Number(drawable.shape.item_width?.value || 10);
    const height = Number(drawable.shape.item_height?.value || 10);
    const radius = Number(drawable.shape.corner_radius?.value || 5);

    if (
        drawable.type === 'shape_drawable' &&
        bg !== 'transparent' &&
        drawable.shape.type === 'rounded_rectangle'
    ) {
        const stl: Partial<DrawableStyle> = {};

        const borderColor = drawable.stroke?.color ? correctColor(drawable.stroke.color) : '';
        const borderWidth = drawable.stroke?.width ? Number(drawable.stroke.width || 1) : '';

        stl.width = width;
        stl.height = height;
        stl.borderRadius = radius;
        stl.background = bg;

        if (borderColor && borderWidth) {
            stl.boxShadow = `inset 0 0 0 ${pxToEm(borderWidth)} ${borderColor}`;
        }

        return stl as DrawableStyle;
    }

    return defaultValue;
}
