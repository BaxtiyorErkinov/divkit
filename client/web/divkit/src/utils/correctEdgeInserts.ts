import type { EdgeInsets } from '../types/edgeInserts';
import { isPositiveNumber } from './isPositiveNumber';
import { edgeInsertsToCss } from './edgeInsertsToCss';
import { isNonNegativeNumber } from './isNonNegativeNumber';

export function correctEdgeInserts(edgeInsets: EdgeInsets | undefined, defaultVal: string): string {
    if (!edgeInsets) {
        return defaultVal;
    }

    const list = [
        edgeInsets.top,
        edgeInsets.right,
        edgeInsets.bottom,
        edgeInsets.left
    ];

    for (let i = 0; i < list.length; ++i) {
        if (list[i] && !isNonNegativeNumber(list[i])) {
            return defaultVal;
        }
    }

    return edgeInsertsToCss(edgeInsets);
}
