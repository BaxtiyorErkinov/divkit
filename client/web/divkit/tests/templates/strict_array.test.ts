import { applyTemplate } from '../../src/utils/applyTemplate';

describe('strict_array', () => {
    test('test_referenced_strict_array_with_non_object_item', () => {
        const json = require('../../../../../test_data/template_test_data/strict_array/test_referenced_strict_array_with_non_object_item.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_referenced_strict_array_with_non_resolvable_item', () => {
        const json = require('../../../../../test_data/template_test_data/strict_array/test_referenced_strict_array_with_non_resolvable_item.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_strict_array_happy_case', () => {
        const json = require('../../../../../test_data/template_test_data/strict_array/test_strict_array_happy_case.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates || {}, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_strict_array_happy_case_referenced', () => {
        const json = require('../../../../../test_data/template_test_data/strict_array/test_strict_array_happy_case_referenced.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError).not.toBeCalled();
    });

    test('test_strict_array_with_non_object_item', () => {
        const json = require('../../../../../test_data/template_test_data/strict_array/test_strict_array_with_non_object_item.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates || {}, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_strict_array_with_non_object_item_ref', () => {
        const json = require('../../../../../test_data/template_test_data/strict_array/test_strict_array_with_non_object_item_ref.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_strict_array_with_non_resolvable_item', () => {
        const json = require('../../../../../test_data/template_test_data/strict_array/test_strict_array_with_non_resolvable_item.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates || {}, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });

    test('test_strict_array_with_non_resolvable_item_ref', () => {
        const json = require('../../../../../test_data/template_test_data/strict_array/test_strict_array_with_non_resolvable_item_ref.json');
        const logError = jest.fn();

        expect(applyTemplate(json.entity, {}, json.templates, logError)).toMatchSnapshot();
        expect(logError.mock.calls).toMatchSnapshot();
    });
});
