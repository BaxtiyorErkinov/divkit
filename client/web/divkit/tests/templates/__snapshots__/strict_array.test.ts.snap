// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`strict_array test_referenced_strict_array_with_non_object_item 1`] = `
Object {
  "json": Object {
    "array": Array [
      0,
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link": Array [
      0,
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_strict_array",
  },
  "templateContext": Object {
    "link": Array [
      0,
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
  },
}
`;

exports[`strict_array test_referenced_strict_array_with_non_resolvable_item 1`] = `
Object {
  "json": Object {
    "array": Array [
      Object {
        "type": "entity_with_complex_property",
      },
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link": Array [
      Object {
        "type": "entity_with_complex_property",
      },
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_strict_array",
  },
  "templateContext": Object {
    "link": Array [
      Object {
        "type": "entity_with_complex_property",
      },
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
  },
}
`;

exports[`strict_array test_strict_array_happy_case 1`] = `
Object {
  "json": Object {
    "array": Array [
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_strict_array",
  },
  "templateContext": Object {},
}
`;

exports[`strict_array test_strict_array_happy_case 2`] = `
Array [
  Array [
    [Error: No such template],
  ],
]
`;

exports[`strict_array test_strict_array_happy_case_referenced 1`] = `
Object {
  "json": Object {
    "array": Array [
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "link": Array [
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_strict_array",
  },
  "templateContext": Object {
    "link": Array [
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
  },
}
`;

exports[`strict_array test_strict_array_with_non_object_item 1`] = `
Object {
  "json": Object {
    "array": Array [
      0,
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_strict_array",
  },
  "templateContext": Object {},
}
`;

exports[`strict_array test_strict_array_with_non_object_item 2`] = `
Array [
  Array [
    [Error: No such template],
  ],
]
`;

exports[`strict_array test_strict_array_with_non_object_item_ref 1`] = `
Object {
  "json": Object {
    "array": Array [
      0,
      Object {
        "link": "https://ya.ru",
        "type": "t_0",
      },
    ],
    "type": "entity_with_strict_array",
  },
  "templateContext": Object {},
}
`;

exports[`strict_array test_strict_array_with_non_object_item_ref 2`] = `
Array [
  Array [
    [Error: No such template],
  ],
]
`;

exports[`strict_array test_strict_array_with_non_resolvable_item 1`] = `
Object {
  "json": Object {
    "array": Array [
      Object {
        "type": "entity_with_complex_property",
      },
      Object {
        "property": Object {
          "value": "https://ya.ru",
        },
        "type": "entity_with_complex_property",
      },
    ],
    "type": "entity_with_strict_array",
  },
  "templateContext": Object {},
}
`;

exports[`strict_array test_strict_array_with_non_resolvable_item 2`] = `
Array [
  Array [
    [Error: No such template],
  ],
]
`;

exports[`strict_array test_strict_array_with_non_resolvable_item_ref 1`] = `
Object {
  "json": Object {
    "array": Array [
      Object {
        "type": "t_0",
      },
      Object {
        "link": "https://ya.ru",
        "type": "t_0",
      },
    ],
    "type": "entity_with_strict_array",
  },
  "templateContext": Object {},
}
`;

exports[`strict_array test_strict_array_with_non_resolvable_item_ref 2`] = `
Array [
  Array [
    [Error: No such template],
  ],
]
`;
